File: omf-core-framework/src/main/java/com/samares_engineering/omf/omf_core_framework/utils/LockerManager.java
/*******************************************************************************
 * @copyright Copyright (c) 2022-2023 Samares-Engineering
 * @Licence: EPL 2.0
 * @Author:   Quentin Cespédès, Clément Mezerette, Hugo Stinson
 * @since     0.0.0
 ******************************************************************************/
package com.samares_engineering.omf.omf_core_framework.utils;

import com.nomagic.magicdraw.core.Project;
import com.nomagic.magicdraw.core.ProjectUtilities;
import com.nomagic.magicdraw.teamwork2.locks.ILockProjectService;
import com.nomagic.magicdraw.teamwork2.locks.LockInfo;
import com.nomagic.magicdraw.teamwork2.locks.LockService;
import com.nomagic.uml2.ext.jmi.helpers.ModelHelper;
import com.nomagic.uml2.ext.magicdraw.classes.mdkernel.Element;
import com.nomagic.uml2.ext.magicdraw.classes.mdkernel.ElementTaggedValue;
import com.nomagic.uml2.ext.magicdraw.classes.mdkernel.TaggedValue;
import com.nomagic.uml2.ext.magicdraw.compositestructures.mdinternalstructures.ConnectorEnd;
import com.nomagic.uml2.ext.magicdraw.mdprofiles.Stereotype;
import com.nomagic.uml2.transaction.ModelValidationResult;
import com.samares_engineering.omf.omf_core_framework.errors.OMFErrorHandler;
import com.samares_engineering.omf.omf_core_framework.errors.exceptions.general.OMFLockException;

import javax.annotation.CheckForNull;
import java.beans.PropertyChangeEvent;
import java.util.*;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class LockerManager {

    private Project project;
    private ILockProjectService projectService = null;
    private static LockerManager instance;

    private LockerManager(Project project){
        this.projectService = LockService.getLockService(project);
        this.project = project;
    }

    public static LockerManager getInstance(){
        return getInstance(OMFUtils.getProject());
    }

    public static LockerManager getInstance(Project project){
        if(instance == null || instance.project != project)
            instance = new LockerManager(project);

        return instance;
    }

    public boolean isLockedByOther(Collection<Element> listElementsToLock){

        return listElementsToLock.stream().anyMatch(element -> isLockedByMe(element) && isLocked(element));

//        for(Element e: listElementsToLock){
//            if (!projectService.isLockedByMe(e) && projectService.isLocked(e)) {
//                return false;
//            }
//        }
//        return true;
    }

    private boolean isLockedByMe(Element element) {
        if(projectService == null) return false;
        return !projectService.isLockedByMe(element);
    }

    public boolean isEditable(Element element){

        boolean isInAttachedProject = ProjectUtilities.isElementInAttachedProject(element);
        boolean isEditable = element.isEditable();
        return isEditable && !isInAttachedProject;
    }
    public boolean isLockedByOther(Element element) {
        if(projectService == null) return false;
        boolean isLocked = isLockedByMe(element) && isLocked(element);

        return isLocked;
    }

    public boolean isLocked(Element element){
        if(projectService == null) return false;
        return projectService.isLocked(element);
    }
    
    public void checkIfEditable(Element element) {
        boolean isEditable = isEditable(element) && !isLockedByOther(element);
        String lockInfo = "";

        projectService.getLockInfo(element);

        if(isLockedByOther(element)) {
//            throw new OMFLockException("[LOCK ERROR] Element is locked by  " + projectService.getLockInfo(element), element);
            OMFErrorHandler.handleException(new OMFLockException("[LOCK ERROR] Element is locked by  " + lockInfo + "   PLEASE UNDO and resolve this lock issue", element));
            return;
        }

        if(!isEditable) {
//            throw new OMFLockException("[LOCK ERROR] Element is not editable  " + projectService.getLockInfo(element), element);}
            OMFErrorHandler.handleException(new OMFLockException("[LOCK ERROR] Element is not editable => PLEASE UNDO and check if these elements are accessible (shared and read-only projects, libraries etc)  " + lockInfo, element));
            return;
        }
//        if(Objects.isNull(projectService.getLockInfo(element)))
//            return;
        if(!projectService.isLocked(element)) {
            OMFErrorHandler.handleException(new OMFLockException("[LOCK ERROR] Element is not lock. => PLEASE UNDO and lock these element before actions  " + lockInfo, element));
//            projectService.lockElements(Collections.singleton(element), new MyProgressStatus());

            return;
        }


    }


    public OMFLockException checkIfEditable2(Element element) {
        boolean isEditable = isEditable(element) || isLockedByOther(element);
        String lockInfo = "";
        projectService.getLockInfo(element);

        if(isLockedByOther(element)) {
//            throw new OMFLockException("[LOCK ERROR] Element is locked by  " + projectService.getLockInfo(element), element);
            return new OMFLockException("     [LOCKED BY] " + lockInfo, element);
        }

        if(!isEditable) {
//            throw new OMFLockException("[LOCK ERROR] Element is not editable  " + projectService.getLockInfo(element), element);}
            return new OMFLockException("     [NON EDITABLE]  " + lockInfo, element);
        }
//        if(Objects.isNull(projectService.getLockInfo(element)))
//            return;
        if(!projectService.isLocked(element)) {
            return new OMFLockException("     [NOT LOCK] " + lockInfo, element);
//            projectService.lockElements(Collections.singleton(element), new MyProgressStatus());
        }
        return null;
    }


    public List<OMFLockException> checkCreation(@CheckForNull List<PropertyChangeEvent> events, Set<Element> checkedElements) {
        return defaultCheck(events, checkedElements);
    }


    public Collection<OMFLockException> checkUpdate(@CheckForNull List<PropertyChangeEvent> events, Set<Element> checkedElements) {
        return defaultCheck(events, checkedElements);
    }
    public Collection<OMFLockException> checkDelete(@CheckForNull List<PropertyChangeEvent> events, Set<Element> checkedElements) {
        return defaultCheck(events, checkedElements);
    }

    private List<OMFLockException> defaultCheck(List<PropertyChangeEvent> events, Set<Element> checkedElements) {
        Map<Element, String> elementsToCheckMap = filterElementToCheck(events, checkedElements);

        checkedElements.addAll(elementsToCheckMap.keySet());

        ArrayList<OMFLockException> lockExceptions = new ArrayList<>();

        elementsToCheckMap.forEach((element, propertyName) ->
                checkElement(element, propertyName).ifPresent(lockExceptions::add));

        return lockExceptions;
    }

    /**
     * Check if an element is Editable and return a OMFLockException if the element is not editable (e.g. project Usages) or if it is not movable (e.g. locks)
     * @param elementToCheck
     * @param propertyName
     * @return Optional OMFLockException
     */
    private Optional<OMFLockException> checkElement(Element elementToCheck, String propertyName) {
        boolean isEditable = isEditable(elementToCheck);
        boolean isMovable = elementToCheck.getOwner() == null ||  ModelHelper.canMoveChildInto(elementToCheck.getOwner(), elementToCheck);

        LockInfo lockInfo = null;
        if(projectService != null)
            lockInfo = projectService.getLockInfo(elementToCheck);

        String sLockInfo = Objects.isNull(lockInfo)? "":  lockInfo.toString();

        if(!isEditable) return Optional.of(new OMFLockException("     [NON EDITABLE]  " + sLockInfo + " - cause: " + propertyName, elementToCheck));


        if(!isMovable) return Optional.of(new OMFLockException("     [NON MOVABLE]  " + sLockInfo+ " - cause: " + propertyName, elementToCheck));


        return Optional.empty();
    }

    /**
     * Will filter all events removing all already checked elements, and irrelevant elements. It will also retrieve the real modified element in the case of TaggedValue .
     * @param events
     * @param checkedElements
     * @return map<Modified Element, Modified PropertyName>
     */
    private Map<Element, String> filterElementToCheck(@CheckForNull List<PropertyChangeEvent> events, Set<Element> checkedElements) {
        Predicate <? super PropertyChangeEvent> hasTheGoodClass =
                   (evt -> Element.class.isInstance(evt.getSource())
                            && !ConnectorEnd.class.isInstance(evt.getSource())
                            && !Stereotype.class.isInstance(evt.getSource())
                            && !Stereotype.class.isInstance(((Element) evt.getSource()).getOwner()));
        Predicate <? super PropertyChangeEvent> isNotTaggedValue = evt -> !(evt.getPropertyName().equals("_elementTaggedValue"));
        Predicate <? super PropertyChangeEvent> isNotEnd = evt -> !(evt.getPropertyName().equals("end"));
        Predicate <? super PropertyChangeEvent> isNotParticipatesInInteraction = evt -> !(evt.getPropertyName().startsWith("participates"));
        Predicate <? super PropertyChangeEvent> isTaggedValue =        evt -> (evt.getSource() instanceof TaggedValue);
        Predicate <? super PropertyChangeEvent> isElementTaggedValue = evt -> (evt.getSource() instanceof ElementTaggedValue);
        Predicate <? super PropertyChangeEvent> isNotAlreadyChecked = (evt -> !checkedElements.contains(evt.getSource()));


        Predicate <? super Map.Entry> notNull = (entry -> entry.getKey() != null);

        Predicate <? super Object> isNotAComputedProperty = evt -> !(((PropertyChangeEvent) evt).getPropertyName().startsWith("_"));

        Function<? super  PropertyChangeEvent, Element> manageElementTaggedValue = evt -> {
            if(isElementTaggedValue.test(evt)) {
                checkedElements.addAll(((ElementTaggedValue) evt.getSource()).getValue().stream().filter(Element.class::isInstance).collect(Collectors.toList()));
                return null;
            }
            if(isTaggedValue.test(evt)) {
                checkedElements.add((TaggedValue) evt.getSource());
                return null;
            }
            return (Element) evt.getSource();
        };

        Function<? super  PropertyChangeEvent, Map.Entry<Element, String>> evtToEntry = evt -> {
            Element element = manageElementTaggedValue.apply(evt);
            return new AbstractMap.SimpleEntry<>(element, evt.getPropertyName());
        };



        HashMap<Element, String> elementToPropertieNamesMap = new HashMap<>();
        events.stream()
                .filter(isNotTaggedValue)
                .filter(isNotEnd)
                .filter(isNotParticipatesInInteraction)
                .filter(isNotAComputedProperty)
                .filter(hasTheGoodClass)
                .filter(isNotAlreadyChecked)
                .map(evtToEntry)
                .filter(notNull)
                .forEach(entry -> elementToPropertieNamesMap.put(entry.getKey(),entry.getValue()));

        return elementToPropertieNamesMap;
    }

    public ModelValidationResult validateLocks(Element element)  {
        boolean isEditable = isEditable(element) && !isLockedByOther(element);

        if(!isLocked(element)) return new ModelValidationResult(element, "[LOCK ERROR] Element is not locked ");

        if(isLockedByOther(element)) return new ModelValidationResult(element, "[LOCK ERROR] Element is locked by  " + projectService.getLockInfo(element));

        if(!isEditable) return new ModelValidationResult(element, "[LOCK ERROR] Element is not editable  " + projectService.getLockInfo(element));
        return null;

    }

}

File: omf-core-framework/src/main/java/com/samares_engineering/omf/omf_core_framework/utils/LockerManager.kt
/*******************************************************************************
 * @copyright Copyright (c) 2022-2023 Samares-Engineering
 * @Licence: EPL 2.0
 * @Author:   Quentin Cespédès, Clément Mezerette, Hugo Stinson, Calliopé Danton Laloy
 * @since     0.0.0
 */
package com.samares_engineering.omf.omf_core_framework.utils

import com.nomagic.magicdraw.core.Project
import com.nomagic.magicdraw.core.ProjectUtilities
import com.nomagic.magicdraw.teamwork2.locks.ILockProjectService
import com.nomagic.magicdraw.teamwork2.locks.LockInfo
import com.nomagic.magicdraw.teamwork2.locks.LockService
import com.nomagic.uml2.ext.jmi.helpers.ModelHelper
import com.nomagic.uml2.ext.magicdraw.classes.mdkernel.Element
import com.nomagic.uml2.ext.magicdraw.classes.mdkernel.ElementTaggedValue
import com.nomagic.uml2.ext.magicdraw.classes.mdkernel.TaggedValue
import com.nomagic.uml2.ext.magicdraw.compositestructures.mdinternalstructures.ConnectorEnd
import com.nomagic.uml2.ext.magicdraw.mdprofiles.Stereotype
import com.nomagic.uml2.transaction.ModelValidationResult
import com.samares_engineering.omf.omf_core_framework.errors.LegacyErrorHandler
import com.samares_engineering.omf.omf_core_framework.errors.exceptions.general.LockException
import com.samares_engineering.omf.omf_core_framework.utils.twc.TWCUtils
import java.beans.PropertyChangeEvent
import java.util.*
import java.util.function.Function
import java.util.function.Predicate

object LockerManager {
    private var project: Project = defaultProject
    private val areProjectTWC = HashMap<Project, Boolean>()

    private val defaultProject: Project
        get() = OMFUtils.getProject()
    private val projectService: ILockProjectService?
        get() = LockService.getLockService(project)



    @JvmStatic
    fun withProject(project: Project): LockerManager {
        this.project = project
        return this
    }


    fun canEdit(element: Element): Boolean {
        return isEditable(element) &&
                isLockFree(element) &&
                isMovable(element)
    }

    fun isLockFree(element: Element): Boolean {
        return isNotTWCProject() || (isLocked(element) && isLockedByMe(element))
    }


    fun isNotTWCProject(): Boolean =  !(areProjectTWC.computeIfAbsent(project) { TWCUtils.isItTWCProject(project) })



    fun isLockedByOther(listElementsToLock: Collection<Element>): Boolean {
        return listElementsToLock.any { element: Element -> isLockedByMe(element) && isLocked(element) }

        //        for(Element e: listElementsToLock){
//            if (!projectService.isLockedByMe(e) && projectService.isLocked(e)) {
//                return false;
//            }
//        }
//        return true;
    }

    fun isLockedByMe(element: Element): Boolean {
        if (projectService == null) return false
        return !projectService!!.isLockedByMe(element)
    }

    fun isEditable(element: Element): Boolean {
        val isInAttachedProject = ProjectUtilities.isElementInAttachedProject(element)
        val isEditable = element.isEditable
        return isEditable && !isInAttachedProject
    }

    fun isLockedByOther(element: Element): Boolean {
        if (projectService == null) return false
        val isLocked = isLockedByMe(element) && isLocked(element)

        return isLocked
    }

    fun isLocked(element: Element): Boolean {
        if (projectService == null) return false
        return projectService!!.isLocked(element)
    }

    fun checkIfEditable(element: Element) {
        val isEditable = isEditable(element) && !isLockedByOther(element)
        val lockInfo = ""

        projectService!!.getLockInfo(element)

        if (isLockedByOther(element)) {
            LegacyErrorHandler.handleException(
                LockException(
                    "[LOCK ERROR] Element is locked by  $lockInfo   PLEASE UNDO and resolve this lock issue", element
                )
            )
            return
        }

        if (!isEditable) {
            LegacyErrorHandler.handleException(
                LockException(
                    "[LOCK ERROR] Element is not editable => PLEASE UNDO and check if these elements are accessible (shared and read-only projects, libraries etc)  $lockInfo",
                    element
                )
            )
            return
        }

        if (!projectService!!.isLocked(element)) {
            LegacyErrorHandler.handleException(
                LockException(
                    "[LOCK ERROR] Element is not lock. => PLEASE UNDO and lock these element before actions  $lockInfo",
                    element
                )
            )
            return
        }
    }


    fun checkIfEditable2(element: Element): LockException? {
        val isEditable = isEditable(element) || isLockedByOther(element)
        val lockInfo = ""
        projectService!!.getLockInfo(element)

        if (isLockedByOther(element)) {
            return LockException("     [LOCKED BY] $lockInfo", element)
        }

        if (!isEditable) {
            return LockException("     [NON EDITABLE]  $lockInfo", element)
        }

        if (!projectService!!.isLocked(element)) {
            return LockException("     [NOT LOCK] $lockInfo", element)
        }
        return null
    }


    fun checkCreation(events: List<PropertyChangeEvent>, checkedElements: MutableSet<Element>): List<LockException> {
        return defaultCheck(events, checkedElements)
    }


    fun checkUpdate(
        events: List<PropertyChangeEvent>,
        checkedElements: MutableSet<Element>
    ): Collection<LockException> {
        return defaultCheck(events, checkedElements)
    }

    fun checkDelete(
        events: List<PropertyChangeEvent>,
        checkedElements: MutableSet<Element>
    ): Collection<LockException> {
        return defaultCheck(events, checkedElements)
    }

    private fun defaultCheck(
        events: List<PropertyChangeEvent>,
        checkedElements: MutableSet<Element>
    ): List<LockException> {
        val elementsToCheckMap = filterElementToCheck(events, checkedElements)

        checkedElements.addAll(elementsToCheckMap.keys)

        val lockExceptions = ArrayList<LockException>()

        elementsToCheckMap.forEach { (element: Element, propertyName: String) ->
            checkElement(
                element,
                propertyName
            ).ifPresent { e: LockException -> lockExceptions.add(e) }
        }

        return lockExceptions
    }

    /**
     * Check if an element is Editable and return a OMFLockException if the element is not editable (e.g. project Usages) or if it is not movable (e.g. locks)
     * @param elementToCheck
     * @param propertyName
     * @return Optional OMFLockException
     */
    private fun checkElement(elementToCheck: Element, propertyName: String): Optional<LockException> {
        val isEditable = isEditable(elementToCheck)
        val isMovable = isMovable(elementToCheck)

        var lockInfo: LockInfo? = null
        if (projectService != null) lockInfo = projectService!!.getLockInfo(elementToCheck)

        val sLockInfo = if (Objects.isNull(lockInfo)) "" else lockInfo.toString()

        if (!isEditable) return Optional.of(
            LockException(
                "     [NON EDITABLE]  $sLockInfo - cause: $propertyName", elementToCheck
            )
        )


        if (!isMovable) return Optional.of(
            LockException(
                "     [NON MOVABLE]  $sLockInfo - cause: $propertyName", elementToCheck
            )
        )


        return Optional.empty()
    }

     fun isMovable(elementToCheck: Element) =
        elementToCheck.owner == null || ModelHelper.canMoveChildInto(
            elementToCheck.owner, elementToCheck
        )

    /**
     * Will filter all events removing all already checked elements, and irrelevant elements. It will also retrieve the real modified element in the case of TaggedValue .
     * @param events
     * @param checkedElements
     * @return map<Modified Element, Modified PropertyName>
    </Modified> */
    private fun filterElementToCheck(
        events: List<PropertyChangeEvent>,
        checkedElements: MutableSet<Element>
    ): Map<Element, String> {
        val hasTheGoodClass: Predicate<in PropertyChangeEvent> =
            (Predicate { evt: PropertyChangeEvent ->
                (Element::class.java.isInstance(evt.source)
                        && !ConnectorEnd::class.java.isInstance(evt.source)
                        && !Stereotype::class.java.isInstance(evt.source)
                        && !Stereotype::class.java.isInstance((evt.source as Element).owner))
            })
        val isNotTaggedValue: Predicate<in PropertyChangeEvent> =
            Predicate { evt: PropertyChangeEvent -> evt.propertyName != "_elementTaggedValue" }
        val isNotEnd: Predicate<in PropertyChangeEvent> =
            Predicate { evt: PropertyChangeEvent -> evt.propertyName != "end" }
        val isNotParticipatesInInteraction: Predicate<in PropertyChangeEvent> =
            Predicate { evt: PropertyChangeEvent -> !(evt.propertyName.startsWith("participates")) }
        val isTaggedValue: Predicate<in PropertyChangeEvent> =
            Predicate { evt: PropertyChangeEvent -> (evt.source is TaggedValue) }
        val isElementTaggedValue: Predicate<in PropertyChangeEvent> =
            Predicate { evt: PropertyChangeEvent -> (evt.source is ElementTaggedValue) }
        val isNotAlreadyChecked: Predicate<in PropertyChangeEvent> =
            (Predicate { evt: PropertyChangeEvent -> !checkedElements.contains(evt.source) })


        val notNull: Predicate<in Map.Entry<Element?, String>> = (Predicate { entry: Map.Entry<Element?, String> -> entry.key != null })

        val isNotAComputedProperty =
            Predicate { evt: Any -> !((evt as PropertyChangeEvent).propertyName.startsWith("_")) }

        val manageElementTaggedValue: Function<in PropertyChangeEvent, Element?> =
            Function<PropertyChangeEvent, Element?> { evt: PropertyChangeEvent ->
                if (isElementTaggedValue.test(evt)) {
                    checkedElements.addAll(
                        (evt.source as ElementTaggedValue).value
                            .filter { obj: Element? -> Element::class.java.isInstance(obj) })
                    return@Function null
                }
                if (isTaggedValue.test(evt)) {
                    checkedElements.add(evt.source as TaggedValue)
                    return@Function null
                }
                evt.source as Element
            }

        val evtToEntry: Function<in PropertyChangeEvent, Map.Entry<Element?, String>> =
            Function<PropertyChangeEvent, Map.Entry<Element?, String>> { evt: PropertyChangeEvent ->
                val element = manageElementTaggedValue.apply(evt)
                AbstractMap.SimpleEntry(element, evt.propertyName)
            }


        val elementToPropertiesNamesMap = HashMap<Element, String>()
        events
            .asSequence()
            .filter{isNotTaggedValue.test(it)}
            .filter{isNotEnd.test(it)}
            .filter{isNotParticipatesInInteraction.test(it)}
            .filter{isNotAComputedProperty.test(it)}
            .filter{hasTheGoodClass.test(it)}
            .filter{isNotAlreadyChecked.test(it)}
            .map{evtToEntry.apply(it)}
            .filter{notNull.test(it)}
            .toList()
            .forEach { entry: Map.Entry<Element?, String> -> elementToPropertiesNamesMap[entry.key!!] = entry.value }

        return elementToPropertiesNamesMap
    }

    fun validateLocks(element: Element): ModelValidationResult? {
        val isEditable = isEditable(element) && !isLockedByOther(element)

        if (!isLocked(element)) return ModelValidationResult(element, "[LOCK ERROR] Element is not locked ")

        if (isLockedByOther(element)) return ModelValidationResult(
            element,
            "[LOCK ERROR] Element is locked by  " + projectService!!.getLockInfo(element)
        )

        if (!isEditable) return ModelValidationResult(
            element,
            "[LOCK ERROR] Element is not editable  " + projectService!!.getLockInfo(element)
        )
        return null
    }




}
